python/ray/train/tests/test_trainer_restore.py

```diff
@@ -0,0 +1,391 @@
+        str(tmpdir / "data_parallel_restore_test"),
+        train_loop_per_worker=train_fn,
+        train_loop_config=train_loop_config,
+        datasets=datasets,
+    )
+    result = trainer.fit()
+    assert not result.error
+    assert result.metrics["training_iteration"] == 2
+    assert result.metrics["iterations_since_restore"] == 1
+    assert tmpdir / "data_parallel_restore_test" in result.log_dir.parents
+
+
+@pytest.mark.parametrize("trainer_cls", [XGBoostTrainer, LightGBMTrainer])
+def test_gbdt_trainer_restore(ray_start_6_cpus, tmpdir, trainer_cls):
+    """Tests restoring gradient boosted decision tree trainers.
+    Success criteria:
+    - Picks up at the right iteration. 2 before crash. 3 after. 5 total trees.
+    - Results are being logged to the same directory as before.
+    """
+    exp_name = f"{trainer_cls.__name__}_restore_test"
+    datasets = {"train": ray.data.from_items([{"x": x, "y": x + 1} for x in range(32)])}
+
+    trainer = trainer_cls(
+        label_column="y",
+        params={
+            "objective": (
+                "reg:squarederror" if trainer_cls == XGBoostTrainer else "regression"
+            )
+        },
+        datasets=datasets,
+        scaling_config=ScalingConfig(
+            num_workers=2, trainer_resources={"CPU": 0}, resources_per_worker={"CPU": 1}
+        ),
+        run_config=RunConfig(
+            local_dir=str(tmpdir),
+            name=exp_name,
+            checkpoint_config=CheckpointConfig(num_to_keep=1, checkpoint_frequency=1),
+            callbacks=[FailureInjectionCallback(num_iters=2)],
+            # We also use a stopper, since the restored run will go for
+            # another 5 boosting rounds otherwise.
+            stop={"training_iteration": 5},
+        ),
+        num_boost_round=5,
+    )
+    with pytest.raises(TuneError):
+        result = trainer.fit()
+
+    trainer = trainer_cls.restore(str(tmpdir / exp_name), datasets=datasets)
+    result = trainer.fit()
+    assert not result.error
+    assert result.metrics["training_iteration"] == 5
+    assert result.metrics["iterations_since_restore"] == 3
+    assert tmpdir / exp_name in result.log_dir.parents
+
+
+@pytest.mark.parametrize("trainer_cls", [HuggingFaceTrainer])
+def test_trainer_with_init_fn_restore(ray_start_4_cpus, tmpdir, trainer_cls):
+    """Tests restore for data parallel trainers that take in a `train_init` function
+    and config. Success criteria: same as for data parallel trainers."""
+    exp_name = f"{trainer_cls.__name__}_restore_test"
+
+    if trainer_cls == HuggingFaceTrainer:
+        from ray.train.tests.test_huggingface_trainer import (
+            train_function as hf_init,
+            train_df,
+        )
+
+        trainer_init_fn = hf_init
+        trainer_init_config = {"epochs": 5, "save_strategy": "epoch"}
+        datasets = {"train": ray.data.from_pandas(train_df)}
+    # TODO(ml-team): Add MosaicTrainer test after Mosaic checkpointing is supported
```

Comment:
what does this comment mean?

Reply:
MosaicTrainer doesn't perform checkpointing at all right now and needs to be added. So, restore/`resume_from_checkpoint` will not do anything.

Reply:
oh I see. I didn't notice we are testing multiple trainers in a single test.
maybe just separate them, and create 1 restore test for each trainer?

Reply:
I am grouping tests with similar trainers (ex: xgboost/lightgbm, huggingface/mosaic) where only a few things need to be changed, and they will still show up as 2 different tests in the CI. Do you think that's okay?

Reply:
ah sure, sure 👍